File: DailyChallenges/CPP/LeetCode_93.cpp
class Solution {
public:
    vector<string> restoreIpAddresses(string s) {
        int n = s.size();
        vector<string> ips;
        function<void(vector<int>&, int)> dfs = [&](vector<int>& v, int i){
            if (v.size()>4) return;
            if (i==n) { // reach the end
                if (v.size() != 4) return;
                string ip = accumulate(next(v.begin()), v.end(), to_string(v[0]), 
                    [](string a, int b) { return a + '.' + to_string(b);});
                ips.push_back(ip);
                return;
            }
            int digit = s[i] - '0';
            // possible to add another
            if (v.size()<4){
                v.push_back(digit);
                dfs(v, i+1);
                v.pop_back();
            }
            // or add to last one
            int seg = v.back() * 10 + digit;
            if (seg <= 255 && v.back() != 0){
                v.pop_back(); v.push_back(seg);
                dfs(v, i+1);
                v.pop_back(); v.push_back(seg/10);
            }
        };

        vector<int> v = {s[0]-'0'};
        dfs(v, 1);
        
        return ips;
    }
};

File: DailyChallenges/Scala/LeetCode_93.scala
object Solution {
    def restoreIpAddresses(s: String): List[String] = {
        var ans = List.empty[String]
        val n = s.length
        def dfs(l: List[Int], i: Int): Unit = {
            if (l.length > 4) return
            if (i == n){
                if (l.length < 4) return
                ans = l.reverse.mkString(".") :: ans
            }else{
                val digit = s(i).asDigit
                if (l.length < 4)
                    dfs(digit :: l, i+1)
                
                val seg = l.head * 10 + digit
                if (seg <= 255 && l.head != 0)
                    dfs(seg :: l.tail, i+1)
            }
        }
        dfs(s(0).asDigit :: Nil, 1)
        ans
    }
}
